### ISSUE TITLE
Why a lot of rounds needed for a very simple data set? And outcome is very sensitive to eta.

### ISSUE THREAD

**Issue by sihanding** [dmlc/xgboost#2510]
Hi all, 

I have tried to use xgboost to solve regression problem. The data is artificial/made-up and it is really simple:
if gender = "male", then y = 1; if gender = "female", then y = 5. 

However, for such a simple data set, it seems to be difficult for xgboost to capture the pattern for small number of rounds, and it is very sensitive to the choice of eta...

## Environment info
Operating System: Windows 7
Compiler: R version 3.4.0
Package used (python/R/jvm/C++): xgboost(R)
`xgboost` version used: xgboost_0.6-4

## Steps to reproduce

1. **Generate data set**

N = 10
data = data.frame(y = rep(c(1, 5), each = N), gender = rep(c("male", "female"), each = N))

2. **One-hot encoding**

data_ohe = model.matrix(~.-1, data[,which(names(data) != "y")])

3. **Train the model. Since there is only one feature, I use only 2 rounds, with very small eta. In my understanding, one round is simple a tree, and in this simple case, one tree should be sufficient. However, the prediction I got is way-off the target...**

xgb.model = xgboost(data = as.matrix(data_ohe), label = data$y, booster = "gbtree", 
                    objective = "reg:linear", max_depth = 2, nround = 2, eta = 0.001)
predict(xgb.model, newdata = as.matrix(data_ohe))

## My questions

1. I tried to adjust eta and nrounds, and for large eta, say 0.5, it converges quickly to the correct results. 
So my question is: why would xgboost need so many rounds/trees for such a simple data set? And why would the results so sensitive to eta? 
2. In my intuition, nround =1 would suffice and the outcome is in-sensitive to eta for this situation. Or did I misunderstand something.

**Comment by sihanding** [comment]
Here is an update: if I change the data to be:
y = x + 1 for male, and y = 3x + 4 for female. 
See the picture for my data. 
![rplot01](https://user-images.githubusercontent.com/30112714/28118117-2e5dc2f6-6710-11e7-8e06-1ee15fe3ab66.png)

And I fit a xgboost method to it:
xgb.model = xgboost(data = as.matrix(data_ohe), label = data$y, booster = "gblinear", 
                    objective = "reg:linear", max_depth = 10, nround = 100, eta = 0.5)

The fitted outcome (red dots) are: 
![rplot2](https://user-images.githubusercontent.com/30112714/28118161-57a49162-6710-11e7-92f4-eb4669f383da.png)

This basically suggests that xgboost cannot work with interactions between variables, while if I fit the data to a linear regression model with interaction terms, then the fit is perfect.

**Comment by khotilov** [comment]
On your question:  eta (learning rate) is a kind of regularization, and it's not surprising that such simple and clean data would strongly depend on it. The smaller learning rate, the more iterations are necessary to converge. Keep in mind that the gbtree booster by default has non-zero L2 regularization parameter (lambda=1), so set it to zero along with eta=1 if you want to see immediate convergence. Otherwise tree leaf weights would tend to be "shrunk". Also, it is generally useful to set `base_score` to a meaningful value (to mean of the label in this regression case) so that boosting does not spend iterations learning the baseline when using eta<1.

And I don't follow your conclusions in the second part. It is clear from the plot that the tree booster did learn the interactions, while not in a perfectly smooth manner, as is typical for tree learners. The linear booster learns simple linear regression: it's smooth but here it did not include any interactions.

**Comment by sihanding** [comment]
Hi khotilov, 
Thank you for your answer. 

From your answer, I understand that if nrounds>1, then eta will have strong influence on the outcome. I am still not quite sure why it will take more than one tree for such simple data set... 

For the second part, to make it more clear, what my conclusions are the following:
1. "gbtree" boosting method will learn interactions. For example, you see red dots fit almost perfectly with the black dots.
![xgboost_10round_0_5shrinkage](https://user-images.githubusercontent.com/30112714/28206374-e5c45466-6885-11e7-8cb1-f1f39e6b9885.png)

2. "gblinear" boosting method ignores interactions between variables, as you only see two parallel lines for the fitted results with no interactions.

**Comment by khotilov** [comment]
Please see the example below.
```R
library(xgboost)
N = 250
d = data.frame(y = rep(c(1, 5), each = N), f = rep(c(0, 1), each = N), m = rep(c(1, 0), each = N))
base_score = mean(d$y)

bst = xgboost(data = as.matrix(d[,-1]), label = d$y, objective = "reg:linear", nround = 1, 
              eta = 1, lambda = 0, max_depth = 2, base_score = base_score)
table(predict(bst, as.matrix(d[,-1])))
xgb.dump(bst)
base_score - 2
base_score + 2

bst = xgboost(data = as.matrix(d[,-1]), label = d$y, objective = "reg:linear", nround = 1, 
              eta = 0.1, lambda = 0, max_depth = 2, base_score = base_score)
table(predict(bst, as.matrix(d[,-1])))
xgb.dump(bst)
# equivalent to
base_score - 2 * 0.1
base_score + 2 * 0.1
```
Output:
```R
> bst = xgboost(data = as.matrix(d[,-1]), label = d$y, objective = "reg:linear", nround = 1, 
+               eta = 1, lambda = 0, max_depth = 2, base_score = base_score)
[1]	train-rmse:0.000000 
> table(predict(bst, as.matrix(d[,-1])))

  1   5 
250 250 
> xgb.dump(bst)
[1] "booster[0]"                      "0:[f0<0.5] yes=1,no=2,missing=1"
[3] "1:leaf=-2"                       "2:leaf=2"                       
> base_score - 2
[1] 1
> base_score + 2
[1] 5
> 
> bst = xgboost(data = as.matrix(d[,-1]), label = d$y, objective = "reg:linear", nround = 1, 
+               eta = 0.1, lambda = 0, max_depth = 2, base_score = base_score)
[1]	train-rmse:1.800000 
> table(predict(bst, as.matrix(d[,-1])))

2.79999995231628 3.20000004768372 
             250              250 
> xgb.dump(bst)
[1] "booster[0]"                      "0:[f0<0.5] yes=1,no=2,missing=1"
[3] "1:leaf=-0.2"                     "2:leaf=0.2"                     
> # equivalent to
> base_score - 2 * 0.1
[1] 2.8
> base_score + 2 * 0.1
[1] 3.2
```

**Comment by sihanding** [comment]
Hi khotilov,

Right, the base_score is the key! Now I understand it! Thanks a lot for the examples.
